server/datastore/mysql/hosts.go

```diff
@@ -1087,6 +1106,111 @@ func (d *Datastore) CleanupExpiredHosts(ctx context.Context) error {
 	return nil
 }
 
+func (d *Datastore) ListHostDeviceMapping(ctx context.Context, id uint) ([]*fleet.HostDeviceMapping, error) {
+	stmt := `
+    SELECT
+      id,
+      host_id,
+      email,
+      source
+    FROM
+      host_emails
+    WHERE
+      host_id = ?
+    ORDER BY
+      email, source`
+
+	var mappings []*fleet.HostDeviceMapping
+	err := sqlx.SelectContext(ctx, d.reader, &mappings, stmt, id)
+	if err != nil {
+		return nil, ctxerr.Wrap(ctx, err, "select host emails by host id")
+	}
+	return mappings, nil
+}
+
+func (d *Datastore) ReplaceHostDeviceMapping(ctx context.Context, hid uint, mappings []*fleet.HostDeviceMapping) error {
+	for _, m := range mappings {
+		if hid != m.HostID {
+			return ctxerr.Errorf(ctx, "host device mapping are not all for the provided host id %d, found %d", hid, m.HostID)
+		}
+	}
+
+	// the following SQL statements assume a small number of emails reported
+	// per host.
+	const (
+		selStmt = `
+      SELECT
+        id,
+        email,
+        source
+      FROM
+        host_emails
+      WHERE
+        host_id = ?`
+
+		delStmt = `
+      DELETE FROM
+        host_emails
+      WHERE
+        id IN (?)`
+
+		insStmt = `
+      INSERT INTO
+        host_emails (host_id, email, source)
+      VALUES`
+
+		insPart = ` (?, ?, ?),`
+	)
+
+	// index the mappings by email and source, to quickly check which ones
+	// need to be deleted and inserted
+	toIns := make(map[string]*fleet.HostDeviceMapping)
+	for _, m := range mappings {
+		toIns[m.Email+"\n"+m.Source] = m
+	}
+
+	return d.withRetryTxx(ctx, func(tx sqlx.ExtContext) error {
```

Comment:
Random thought, no need to address or even read for that matter, just thinking "out loud": Sometimes this pattern makes me feel like we should be storing 1 row per host and json content rather than one row per entity within that json.

In this case, for search by email it would add the JSON op to the query, but search would be rare so it could be fine, and storage of the data would be very common. And for that, we would insert once and update from that point on, no selects needed.

Again... just a random thought... I had this thought right after thinking "we don't TECHNICALLY need the autoincrement id" which I know is not an ideal thought when modeling data in a relational db, so take this with a couple of grains of salt :)  (BUT! this would solve, because host_id would be the key).

Reply:
yeah for sure it makes storage easier / closer to a document db, but for search I vaguely remember that mysql 5.7 does not support indexing on JSON data and JSON funcs being relatively slow, though I don't know if it would be significant enough to matter. IIRC mysql 8 improves things a lot for json support.